specgen/specgen/hellocap/hellocap.go

```diff
@@ -0,0 +1,57 @@
+// Package hellocap provides facilities for capturing TLS ClientHellos.
+package hellocap
+
+import (
+	"crypto/tls"
+	"net"
+	"net/http"
+)
+
+// Server is a TLS server used to capture incoming ClientHellos. This is built specifically for the
+// specgen use case. Performance is not prioritized and features are intentionally minimal.
+type Server struct {
+	s  http.Server
+	hl *handlerListener
+}
+
+// NewServer creates a new hello-capturing HTTP server. In the provided handler, use FromRequest to
+// obtain the TLS ClientHello provided by the client.
+func NewServer(handler http.Handler, addr string, cert tls.Certificate) (*Server, error) {
+	// handlerListener does the heavy lifting for us, connecting the TCP-level logic with the HTTP
+	// request contexts.
+	hl, err := newHandlerListener(addr, handler)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Server{
+		http.Server{
+			Handler: hl,
+			TLSConfig: &tls.Config{
+				Certificates: []tls.Certificate{cert},
+				// We disable session ticket support to prevent browsers from sending the pre-shared
+				// key extension; this extension is not supported by utls.FingerprintClientHello.
+				SessionTicketsDisabled: true,
+			},
+		},
+		hl,
+	}, nil
+}
+
+// Addr returns the network address this server listens on. Until ListenAndServe is called, behavior
+// is undefined for incoming connections to this address.
+func (s *Server) Addr() net.Addr {
```

Comment:
If you contrive a way to make `Server` embed `net.Listener`, you don't need these methods. And I think it makes a sort of sense:

```
type Server struct {
    // it's an http server...
    http.Server
    // that listens to requests...
    net.Listener
}
```

Reply:
That's part of what I meant by "features are intentionally minimal".

I think we could make that work, but it'd be a significant amount more code / complexity.  I don't see any motivation for this.

Reply:
You mean extra complexity in that you have to handle the non-TLS case comprehensively? Fair enough - I didn't see that when I wrote my original comment. The motivation would be e.g. exposing timeout settings to the caller.